lib/membrane_mp4/muxer/cmaf.ex

```diff
@@ -30,138 +36,210 @@ defmodule Membrane.MP4.Muxer.CMAF do
-    # - old input caps - must be used to generate a segment from the cached samples, have to be cached as the next handle_process call will use new input caps
-    # - new output caps - must follow freshly generated segment from the old input caps
-    {caps, state} =
+    state
+    |> put_in([:pad_to_track, pad], track_data)
+    |> put_in([:samples, pad], [])
+    |> then(&{:ok, &1})
+  end
+
+  @impl true
+  def handle_caps(pad, %Membrane.MP4.Payload{} = caps, ctx, state) do
+    state =
+      update_in(state, [:pad_to_track, pad], fn tr ->
+        track =
+          caps
+          |> Map.from_struct()
+          |> Map.take([:width, :height, :content, :timescale])
+          |> Map.put(:id, tr.id)
+          |> Track.new()
+
+        %{tr | track: track}
+      end)
+
+    if Map.drop(ctx.pads, [:output, pad]) |> Map.values() |> Enum.all?(&(&1.caps != nil)) do
+      caps = generate_caps(state)
+
       cond do
-        # cache caps only if there are cached samples
-        caps != ctx.pads.output.caps and state.samples != [] ->
-          {[], %{state | old_input_caps: ctx.pads.input.caps, new_output_caps: caps}}
+        is_nil(ctx.pads.output.caps) ->
+          {{:ok, caps: {:output, caps}}, state}
 
         caps != ctx.pads.output.caps ->
-          {[caps: {:output, caps}], state}
+          {:ok, %{state | awaiting_caps: {{:update_with_next, pad}, caps}}}
 
         true ->
-          {[], state}
+          {:ok, state}
       end
-
-    {{:ok, caps ++ [redemand: :output]}, state}
+    else
+      {:ok, state}
+    end
   end
 
   @impl true
-  def handle_end_of_stream(:input, ctx, state) do
-    last_dts = hd(state.samples).dts
+  def handle_end_of_stream(Pad.ref(:input, _track_id) = pad, ctx, state) do
+    sample = state.pad_to_track[pad].duration_resolution_queue
+
+    sample_metadata =
+      Map.put(sample.metadata, :duration, hd(state.samples[pad]).metadata.duration)
+
+    sample = %Buffer{sample | metadata: sample_metadata}
+
+    state = update_in(state, [:samples, pad], &[sample | &1])
+
+    if ctx.pads
+       |> Map.drop([:output, pad])
+       |> Map.values()
+       |> Enum.all?(& &1.end_of_stream?) do
+      with {:ok, segment, state} <- Segment.Helper.take_all_samples(state) do
+        {buffer, state} = generate_segment(segment, ctx, state)
+        {{:ok, buffer: {:output, buffer}, end_of_stream: :output}, state}
+      else
+        {:error, :not_enough_data} -> {{:ok, end_of_stream: :output}, state}
+      end
+    else
+      state = put_in(state, [:pad_to_track, pad, :end_timestamp], nil)
 
-    {buffer, state} =
-      generate_segment(ctx.pads.input.caps, %{dts: last_dts, metadata: %{}}, state)
+      {{:ok, redemand: :output}, state}
+    end
+  end
+
+  defp generate_caps(state) do
+    tracks = Enum.map(state.pad_to_track, fn {_pad, track_data} -> track_data.track end)
+
+    header = Header.serialize(tracks)
+
+    content_type =
+      cond do
+        length(tracks) > 1 -> :muxed_audio_video
```

Comment:
Yes, this can be a combination. Actually, I would be a good idea to modify the spec a bit, so that it actually informs about content_type of each track:
```elixir
@type t() :: %Membrane.CMAF.Track{
  content_type: content_type_t() | [content_type_t()]
}
```
So effectively, the fact that it's muxed would be implied by the fact that `content_type` is a list. Then eg. HLS can check if muxed CMAF contains exactly one audio and exactly one video, I don't think we want anything else in this case

Reply:
I saw you already adjusted CMAF Format. Let's release it